js/modules/k6/data/share.go

```diff
@@ -33,103 +31,80 @@ type sharedArray struct {
 	arr []string
 }
 
-func (s sharedArray) wrap(ctxPtr *context.Context, rt *goja.Runtime) goja.Value {
-	cal, err := rt.RunString(arrayWrapperCode)
-	if err != nil {
-		common.Throw(rt, err)
-	}
-	call, _ := goja.AssertFunction(cal)
-	wrapped, err := call(goja.Undefined(), rt.ToValue(common.Bind(rt, s, ctxPtr)))
-	if err != nil {
-		common.Throw(rt, err)
-	}
+type wrappedSharedArray struct {
+	sharedArray
 
-	return wrapped
+	rt       *goja.Runtime
+	freeze   goja.Callable
+	isFrozen goja.Callable
+	parse    goja.Callable
 }
 
-func (s sharedArray) Get(index int) (interface{}, error) {
-	if index < 0 || index >= len(s.arr) {
-		return goja.Undefined(), nil
-	}
-
-	// we specifically use JSON.parse to get the json to an object inside as otherwise we won't be
-	// able to freeze it as goja doesn't let us unless it is a pure goja object and this is the
-	// easiest way to get one.
```

Comment:
I'm still wondering why we stringify the array elements and then parse them back on `Get()`. Attempting to freeze a non-object [would simply return it as is](https://github.com/dop251/goja/blob/f5884268f0c0a0eb94423dcf7c2eda14907375e6/builtin_object.go#L296-L298), so the stringify/parse dance doesn't seem necessary and is probably a lot of overhead.

Reply:
This was more or less explored in https://github.com/dop251/goja/issues/232 and the answer is that ... while for extremely simple values such as String and numbers it might be feasible ... for even Objects, it is likely that sharing them between goja Runtimes will lead to race conditions *even* if frozen. 

Again the EcmaScript standard is meant for *single-threaded* operations ... everything in the standard (except the few Atomic and *Shared* stuff inside of it) is meant for only 1 thing operating on objects from it. So even if theoretically today goja could support it, tomorrow due to change that might break. 

Additionally, if we don't serialize and then deserialize, we will need to actually check that things such as lambdas(which can reference outside objects inside of them) are not present. Proxies will also be problematic, as well as probably property methods and so on ... All of those things are not an issue as long as we serialize to JSON and back, because goja follows the EcmaScript standard and it actually specifically says that things like that don't get serialized :tada: .

Again the "overhead" is minimal and IMO the possible complications far outweigh the amount of additional reasoning and code that will be needed for this to work. Also again ... telling people that their data will be serialized to JSON and back is a thing that they can reason about as well, whatever else we come up with ... likely won't be so easy to reason about and test with available tools.

Reply:
Thanks for explaining, that makes sense. I think I glossed over the initial discussion in that issue. It might be good to link to it in a comment as a future reference for why this is done.